### PR TITLE
Add Lowest Common Ancestor problem for Trees session

### PR DESCRIPTION
**Pull request by etcadinfinitum**

### CODE REVIEW

File: 2020-02-Spring/3_trees/README.md [hunk 1]
```diff
@@ -23,21 +23,111 @@ Problems and solutions for Trees session on May 8, 2020.
 
 <a name="p1"/>
 
-### 1. PROBLEM 1 TODO :bug:
+### 1. Lowest Common Ancestor
```
Comment (hollowsunsets) [inline 2020-02-Spring/3_trees/README.md:26]: This is a very well written solution, IMO! :fire: :tada: 

Just had a few tiny nitpicky suggestions! Feel free to ignore as they are not critically important.

File: 2020-02-Spring/3_trees/common_ancestor/driver.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+def make_test_1():
+    '''
+    Tree:
+            m
+          l   p
+        d    n  x
+      b   f    t z
+     a c e h
+    '''
+    vals = ['m',
```
[the remark this answers is not in the document]
Comment (etcadinfinitum) [reply 2020-02-Spring/3_trees/common_ancestor/driver.py:37]: Not a lot of work, actually. It helps that I've done this problem several times before in different contexts. Building a tree by hand for tests is a PITA though

File: 2020-02-Spring/3_trees/common_ancestor/utils.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
+# Definition for a binary tree node.
+class TreeNode(object):
```
[the remark this answers is not in the document]
Comment (etcadinfinitum) [reply 2020-02-Spring/3_trees/common_ancestor/utils.py:2]: Will do. Did a copy :spaghetti: from the LC problem

File: 2020-02-Spring/3_trees/README.md [hunk 4]
```diff
@@ -94,21 +184,84 @@ Go to [Solution](#s3)   [Top](#top)
...
+Using this information, we can formulate a traversal strategy that 
+harnesses these properties. At each visited node, we test for the 
+following properties:
+
+1. Is the root `null`? If so, we return `null`.
+2. Is the current root equal to one of the target nodes? If so, the 
+   current root must be the LCA of both of the target nodes. We return 
+   the current root.
+3. Is the root's value **greater** than both of the target nodes? If so, 
+   both nodes must be in the _left-hand side_ of the tree. So, we call
```
[the remark this answers is not in the document]
Comment (etcadinfinitum) [reply 2020-02-Spring/3_trees/README.md:208]: Good call

File: 2020-02-Spring/3_trees/common_ancestor/utils.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    def __init__(self, x):
+        self.val = x
+        self.left = None
+        self.right = None
+
+# Reads an input array, building the tree
+# left-to-right, top-to-bottom.
+# A la https://leetcode.com/problems/serialize-and-deserialize-binary-tree/
+def build_tree(values):
```
[the remark this answers is not in the document]
Comment (etcadinfinitum) [reply 2020-02-Spring/3_trees/common_ancestor/utils.py:11]: crowaii

File: 2020-02-Spring/3_trees/README.md (hunk 1 above)
Comment (etcadinfinitum) [reply 2020-02-Spring/3_trees/README.md:26]: Thank you! Your suggestions are good stuff, will make some edits along those lines

File: 2020-02-Spring/3_trees/README.md [hunk 6]
```diff
@@ -23,21 +23,111 @@ Problems and solutions for Trees session on May 8, 2020.
...
 
 #### Scenario
 
-Problem Statement TODO :bug:
+Given a binary search tree (BST), find the lowest common ancestor (LCA) 
+of two given nodes in the BST.
+
+According to the [definition of LCA on Wikipedia](https://en.wikipedia.org/wiki/Lowest_common_ancestor): 
+"The lowest common ancestor is defined between two nodes p and q as 
+the lowest node in T that has both p and q as descendants (where we
```
[the remark this answers is not in the document]
Comment (etcadinfinitum) [reply 2020-02-Spring/3_trees/README.md:37]: This was another straight copy :spaghetti: from LC -- will revise the definition for clarity. You suggestion is good stuff :sunglasses:

File: 2020-02-Spring/3_trees/README.md [hunk 7]
```diff
@@ -94,21 +184,84 @@ Go to [Solution](#s3)   [Top](#top)
...
+
+    # otherwise, the nodes are distributed across the 
+    # two subtrees of the current node. this implies 
+    # that the current node is the LCA.
+    return root
+```
+
+</details>
 
 #### Testing The Solutions OR Driver For Solution
```
Comment (anishprasad01) [inline 2020-02-Spring/3_trees/README.md:254]:
```suggestion
#### Testing The Solutions
```

File: 2020-02-Spring/3_trees/README.md [hunk 8]
```diff
@@ -94,21 +184,84 @@ Go to [Solution](#s3)   [Top](#top)
...
 <a name="s1"/>
 
-### 1. SOLUTION 1 TODO :bug:
+### 1. Lowest Common Ancestor
 
-Source: TODO :bug:
+Source: [LeetCode](https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-search-tree/)
 
-#### Naive/Simple Solution
+#### Solution
```
[the remark this answers is not in the document]
Comment (huyn8) [reply 2020-02-Spring/3_trees/README.md:191]: Second this.

**Comment by huyn8** [comment]
LGTM. I couldn't find anything that has not been addressed by Isabel and Anish. Well done.

File: 2020-02-Spring/3_trees/README.md [hunk 9]
```diff
@@ -94,21 +184,84 @@ Go to [Solution](#s3)   [Top](#top)
...
 
-TODO :bug:
+In this problem, there are a few crucial axioms that help simplify
```
[the remark this answers is not in the document]
Comment (etcadinfinitum) [reply 2020-02-Spring/3_trees/README.md:193]: Fixed in d0d4eeb68c11c1569ad5810b3f75e94bd9c5936b

File: 2020-02-Spring/3_trees/README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (etcadinfinitum) [reply 2020-02-Spring/3_trees/README.md:191]: Fixed in Fixed in d0d4eeb68c11c1569ad5810b3f75e94bd9c5936b

File: 2020-02-Spring/3_trees/common_ancestor/utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (etcadinfinitum) [reply 2020-02-Spring/3_trees/common_ancestor/utils.py:2]: Fixed in 4618d5d32a0d4a295cabf6da414baf47de672fe6

File: 2020-02-Spring/3_trees/README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (etcadinfinitum) [reply 2020-02-Spring/3_trees/README.md:208]: Fixed in 92fb32f8e9a7ad76904752dcd7d2ec9b90df30b6

File: 2020-02-Spring/3_trees/README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (etcadinfinitum) [reply 2020-02-Spring/3_trees/README.md:193]: I'm going to leave "axiom" as is -- I think it is important to use technical terminology for propositions or ground truths in algorithm explanations. If this section had a more academic tone, I would change it, but I think the balance here is fine.

File: 2020-02-Spring/3_trees/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (etcadinfinitum) [reply 2020-02-Spring/3_trees/README.md:37]: Fixed in b8ca94b59e310a2fb318ae13603da86185cae05b

**Pull request merged**